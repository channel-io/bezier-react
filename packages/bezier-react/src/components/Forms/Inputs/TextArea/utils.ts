import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
