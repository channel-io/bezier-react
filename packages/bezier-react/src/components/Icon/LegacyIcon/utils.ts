/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
