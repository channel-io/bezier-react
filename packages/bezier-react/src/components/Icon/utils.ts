import { type SVGProps } from 'react'

const BEZIER_ICON_SYMBOL = Symbol('bezier.icon')

export type IconSource = (props: SVGProps<SVGSVGElement>) => JSX.Element

export type BezierIcon = IconSource & { [BEZIER_ICON_SYMBOL]: true }

export function isBezierIcon(arg: unknown): arg is BezierIcon {
  return (
    typeof arg === 'function' &&
    arg[BEZIER_ICON_SYMBOL] === true
  )
}

export function createBezierIcon(source: IconSource): BezierIcon {
  source[BEZIER_ICON_SYMBOL] = true
  return source as BezierIcon
}
