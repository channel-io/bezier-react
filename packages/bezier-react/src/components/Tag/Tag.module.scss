.TagDeleteIcon {
  cursor: pointer;

  &:where(:focus-visible) {
    outline: 3px solid var(--bgtxt-cobalt-light);
  }
}
