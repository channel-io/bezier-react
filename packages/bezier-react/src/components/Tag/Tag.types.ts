import {
  type BezierComponentProps,
  type ChildrenProps,
  type SizeProps,
  type VariantProps,
} from '~/src/types/props'

import {
  type BaseTagBadgeSize,
  type BaseTagBadgeVariant,
} from '~/src/components/BaseTagBadge'

export type TagSize = BaseTagBadgeSize

export type TagVariant = BaseTagBadgeVariant

interface TagOwnProps {
  /**
   * Handler to be called when the delete button is clicked.
   *
   * The delete button will be not shown if this prop is not provided.
   */
  onDelete?: React.MouseEventHandler
}

export interface TagProps
  extends BezierComponentProps<'div'>,
    ChildrenProps,
    SizeProps<TagSize>,
    VariantProps<TagVariant>,
    TagOwnProps {}
