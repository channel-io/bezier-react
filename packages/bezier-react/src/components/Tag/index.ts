export { Tag } from './Tag'
export type { TagProps, TagSize, TagVariant } from './Tag.types'
