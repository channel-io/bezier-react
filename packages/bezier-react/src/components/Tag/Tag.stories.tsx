import React from 'react'

import {
  type Meta,
  type StoryFn,
  type StoryObj,
} from '@storybook/react'

import {
  TagBadgeSize,
  TagBadgeVariant,
} from '~/src/components/TagBadgeCommon'

import { Tag } from './Tag'

const meta: Meta<typeof Tag> = {
  component: Tag,
  argTypes: {
    size: {
      control: {
        type: 'radio',
      },
      options: TagBadgeSize,
    },
    variant: {
      control: {
        type: 'radio',
      },
      options: TagBadgeVariant,
    },
  },
}
export default meta

const Template: StoryFn<typeof Tag> = ({ children, ...otherProps }) => (
  <Tag {...otherProps}>{ children }</Tag>
)

export const Primary: StoryObj<typeof Tag> = {
  render: Template,

  args: {
    children: 'Design',
    size: TagBadgeSize.M,
    variant: TagBadgeVariant.Default,
    // eslint-disable-next-line no-console
    onDelete: console.log,
    // eslint-disable-next-line no-console
    onClick: console.log,
  },
}
