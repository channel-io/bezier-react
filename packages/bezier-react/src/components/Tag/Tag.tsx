import React, {
  type CSSProperties,
  forwardRef,
  memo,
} from 'react'

import { CancelSmallIcon } from '@channel.io/bezier-icons'
import classNames from 'classnames'

import { type SemanticColor } from '~/src/types/tokens'
import { cssVar } from '~/src/utils/style'
import {
  isEmpty,
  isNil,
} from '~/src/utils/type'

import {
  Icon,
  IconSize,
} from '~/src/components/Icon'
import {
  TagBadgeSize,
  TagBadgeVariant,
  getProperTagBadgeBgColor,
  getProperTagBadgeTypo,
} from '~/src/components/TagBadgeCommon'
import commonStyles from '~/src/components/TagBadgeCommon/TagBadge.module.scss'
import { Text } from '~/src/components/Text'

import { type TagProps } from './Tag.types'

import styles from './Tag.module.scss'

export const TAG_TEST_ID = 'bezier-tag'
export const TAG_DELETE_TEST_ID = 'bezier-tag-delete-icon'

/**
 * `Tag` is a component for representing tag, which shows close icon when `onDelete` property is specified.
 * @example
 * ```tsx
 * <Tag
 *   size={TagBadgeSize.M}
 *   variant={TagBadgeVariant.Default}
 *   onDelete={handleDelete}
 * >
 *   Payment
 * </Tag>
 * ```
 */
export const Tag = memo(forwardRef<HTMLDivElement, TagProps>(function Tag({
  size = TagBadgeSize.M,
  variant = TagBadgeVariant.Default,
  color: givenColor,
  children,
  className,
  onDelete,
  style,
  ...rest
}, forwardedRef) {
  const bgColor: SemanticColor = givenColor || getProperTagBadgeBgColor(variant)

  return (
    <div
      style={{
        '--b-tag-badge-background-color': cssVar(bgColor),
        ...style,
      } as CSSProperties}
      className={classNames(
        styles.Tag,
        commonStyles.TagBadge,
        commonStyles[`size-${size}`],
        className,
      )}
      ref={forwardedRef}
      data-testid={TAG_TEST_ID}
      {...rest}
    >
      { !isEmpty(children) && (
        <Text
          typo={getProperTagBadgeTypo(size)}
          marginHorizontal={2}
          color="txt-black-darkest"
        >
          { children }
        </Text>
      ) }

      { !isNil(onDelete) && (
        <Icon
          role="button"
          tabIndex={0}
          className={styles.CloseIcon}
          source={CancelSmallIcon}
          size={IconSize.XS}
          color="txt-black-darker"
          onClick={(e) => {
            e.stopPropagation()
            onDelete(e)
          }}
          data-testid={TAG_DELETE_TEST_ID}
        />
      ) }
    </div>
  )
}))
