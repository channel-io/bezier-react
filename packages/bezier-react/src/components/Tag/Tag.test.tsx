import React from 'react'

import userEvent from '@testing-library/user-event'

import { render } from '~/src/utils/test'

import {
  TAG_DELETE_TEST_ID,
  TAG_TEST_ID,
  Tag,
} from './Tag'
import { type TagProps } from './Tag.types'

describe('Tag test >', () => {
  const user = userEvent.setup()
  let props: TagProps

  beforeEach(() => {
    props = {}
  })

  const renderTag = (optionProps?: TagProps) => render(<Tag {...props} {...optionProps} />)

  it('Snapshot >', () => {
    const { getByTestId } = renderTag()
    const tag = getByTestId(TAG_TEST_ID)

    expect(tag).toMatchSnapshot()
  })

  describe('Click Event Test >', () => {
    it('onClick function should be called when the tag is clicked', async () => {
      const onClickFn = jest.fn()
      const { getByTestId } = renderTag({ onClick: onClickFn })

      await user.click(getByTestId(TAG_TEST_ID))

      expect(onClickFn).toHaveBeenCalled()
    })

    it('onDelete function should be called when the delete icon is clicked', async () => {
      const onClickFn = jest.fn()
      const onDeleteFn = jest.fn()
      const { getByTestId } = renderTag({ onClick: onClickFn, onDelete: onDeleteFn })

      await user.click(getByTestId(TAG_DELETE_TEST_ID))

      expect(onDeleteFn).toHaveBeenCalled()
    })

    it('onClick function should not be called when the delete icon is clicked', async () => {
      const onClickFn = jest.fn()
      const onDeleteFn = jest.fn()
      const { getByTestId } = renderTag({ onClick: onClickFn, onDelete: onDeleteFn })

      await user.click(getByTestId(TAG_DELETE_TEST_ID))

      expect(onClickFn).not.toHaveBeenCalled()
    })
  })
})
