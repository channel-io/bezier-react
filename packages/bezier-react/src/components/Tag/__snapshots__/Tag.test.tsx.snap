// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag test > Snapshot > 1`] = `
<div
  class="BaseTagBadge size-m variant-default"
  data-testid="bezier-tag"
/>
`;
